File: java/Sorts/src/MergeSort.java
import java.util.ArrayList;
import java.util.List;

/**
 * Merge sort works by copying the source array into two different arrays/vectors/lists,
 * then merging those back into the source array in ordered fashion.
 * First the call stack is setup so that Merge works on the smallest pieces first, then
 * larger.O(NLogN) 
 * @author bradley
 *
 */
class MergeSort {

	private static <T extends Comparable<T>> void merge(T[] source, int left, int mid, int right){
		int i, j, k=left;
		List<T> lower = new ArrayList<>();
		List<T> upper = new ArrayList<>();
		
		for(i=left; i <= mid; i++) {
			lower.add(source[i]);
		}
		
		for(i=mid + 1; i <= right; i++) {
			upper.add(source[i]);
		}
		
		int n = lower.size(), m = upper.size();
		
		i=0; j=0;
		while(i < n && j < m) {
			T lowerElem = lower.get(i);
			T upperElem = upper.get(j);
			
			if(lowerElem.compareTo(upperElem) <= 0) {
				source[k++] = lowerElem;
				i++;
			}
			else {
				source[k++] = upperElem;
				j++;
			}
		}
		
		while(i < n) {
			source[k++] = lower.get(i++); 
		}
		
		while(j < m) {
			source[k++] = upper.get(j++); 
		}
		
	}
	
	static <T extends Comparable<T>> void sort(T[] source, int left, int right) {
		if(left < right) {
			int mid = (left + right) / 2;
			sort(source, left, mid);
			sort(source, mid + 1, right);
			merge(source, left, mid, right);
		}
	}
}

File: csharp/Sorts/MergeSort.cs
using System;
using System.Collections.Generic;

namespace Sorts
{
	/// <summary>
	/// Merge sort works by copying the source array into two different arrays/vectors/lists,
	/// then merging those back into the source array in ordered fashion.
	/// First the call stack is setup so that Merge works on the smallest pieces first, then
	/// larger.O(NLogN) 
	/// </summary>
	public class MergeSort
	{
		private static void Merge<T>(T[] array, int p, int q, int r) where T : IComparable
		{
			int k = p, i, j;
			List<T> lowerHalf = new List<T> ();
			List<T> upperHalf = new List<T> ();

			for (i=p; i <= q; i++) {
				lowerHalf.Add (array [i]);
			}
			for (i=q+1; i <= r; i++) {
				upperHalf.Add (array [i]);
			}

			i = 0; j = 0;

			while (i < lowerHalf.Count && j < upperHalf.Count) {
				if (lowerHalf [i].CompareTo (upperHalf [j]) <= 0) {
					array [k++] = lowerHalf [i++];
				} else {
					array [k++] = upperHalf [j++];
				}
			}

			while (i < lowerHalf.Count) {
				array [k++] = lowerHalf [i++];
			}

			while (j < upperHalf.Count) {
				array [k++] = upperHalf [j++];
			}
		}

		public static void Sort<T>(T[] array, int p, int r) where T : IComparable
		{
			if (p < r) {
				int q = (p + r) / 2;
				Sort (array, p, q);
				Sort (array, q + 1, r);
				Merge (array, p, q, r);
			}
		}
	}
}

